lib/http_server.ex

```diff
@@ -15,19 +18,87 @@ defmodule HTTPServer do
   end
 
   defp serve(socket) do
+    request =
+      socket
+      |> read_line()
+      |> parse()
+
+    response =
+      request
+      |> build_response()
+      |> text()
+
     socket
-    |> read_line()
-    |> write_line(socket)
+    |> write_line(response)
     |> serve()
   end
 
   defp read_line(socket) do
-    {:ok, data} = :gen_tcp.recv(socket, 0)
-    data
+    case :gen_tcp.recv(socket, 0) do
+      {:ok, data} ->
+        data
+
+      {:error, :econnrefused} ->
+        exit(:normal)
+
+      {:error, :closed} ->
+        exit(:normal)
+    end
   end
 
-  defp write_line(data, socket) do
+  defp write_line(socket, data) do
     :gen_tcp.send(socket, data)
     socket
   end
+
+  def parse(message) do
+    [request_data | body] = message |> String.split("\r\n\r\n")
+    [first | headers] = request_data |> String.split("\r\n")
+    [method, path, resource] = first |> String.split(" ")
+
+    req = %Request{
+      method: method,
+      path: path,
+      resource: resource,
+      headers: format_headers(headers, %{}),
+      body: hd(body)
+    }
+
+    case method do
+      "" ->
+        hd(body)
+
+      "POST" ->
+        %{req | method: :post}
+    end
+  end
+
+  def build_response(req) do
+    %Response{
+      status_code: 200,
+      status_message: :ok,
+      resource: req.resource,
+      headers: req.headers,
+      body: req.body
+    }
+  end
+
+  def text(res) do
+    "#{res.resource} #{res.status_code} OK\r\n" <>
```

Comment:
I see a lot of hard-coded "\r\n" around here - maybe time to extract it into a variable?

Reply:
I like how most of this is extracted - I wonder if there's a way you could tie the status_code to the word that follows? (e.g. 200 => OK, 201 => Created, etc.)